### PR TITLE
[MoveosStd] Refactor AccountStorage

### LINKED ISSUE #1067: [Framework] Refactor AccountStorage
Migrate `account_storage` global functions to `context`, and make `account_storage`to a pure module about `AccountStorage`.

1. `account_storage::global_borrow<T>(&Context, address): &T` -> `context::global_borrow<T>(&Context,address): &T`
2. `account_storage::global_borrow_mut<T>(&mut Context, address): &mut T` -> `context::global_borrow_mut<T>(&mut Context, address): &mut T`
3. `account_storage::global_move_to<T>(&mut Context, signer, T)` -> `context::global_move_to<T>(&mut Context, signer, T)`
4. `account_storage::global_move_from<T>(&mut Context, address):T` -> `context::global_move_from<T>(&mut Context, address):T`
5. `account_storage::global_exists<T>(&Context, address):bool` -> `context::global_exists<T>(&Context,address):bool`

Maybe we should rename the `global_borrow`, `global_borrow_mut`, `global_move_to`, `global_move_from` to `borrow_from_account`, `borrow_mut_from_account`, `move_to_account`, `move_from_account` ?

Then, we make the `account_storage` to object style.

```move

module moveos_std::account_storage{

   public fun borrow_mut<T>(self: &mut AccountStorage): &mut T;
}

```

```move
module my_project:my_module{

   public fun my_entry(account_storage: &mut Object<AccountStorage>){
   }
}

```

### LINKED ISSUE #1112: [MoveosStd] Introduce BoundObject
Now, we need a new Object type:

1. The Object belongs to a user.
2. The user cannot transfer the Object or embed it in another struct.

### Application Scenario:
1. AccountStorage, the Object<AccountStorage> can not be transferred or be embedded in another struct.
2. User's default CoinStore for each Coin.
3. Soul binding NFT.

I want to introduce a new flag: bound, BoundObject. It means the Object is bound to the current user. The user can get the `&mut Object<T>`, but cannot take the `Object<T>` directly.

### PR DESCRIPTION
**Pull request by jolestar**
## Summary

Refactor AccountStorage, resolve #1067 

1. Migrate `Context` relative functions from `account_storage` to `context` and rename.
2. Make `account_storage` a pure module about `AccountStorage`.
3. Rename `context::exist_object` to `context:: exists_object`.
4. Make `take_object<T:key>` to `take_object<T:key+store>` and provide `take_object_extend<T:key>`, resolve #1112
5. Remove dependency on `Context` from the `table` and `type_table` modules and provide `context::new_table` and `context::new_type_table` functions.  

### Function names mapping

account::global_borrow => context::borrow_resource
account::global_borrow_mut => context::borrow_mut_resource
account::global_move_to => context::move_resource_to
account::global_move_from => context::move_resource_from
account::global_exists => context::exists_resource

### Motivation

1. Aggregate all storage relative functions to `context` module, and when Move supports self syntax, the developer can directly call the function on `ctx`, like `ctx.borrow_resource()`.
2. Provide Object style AccountStorage programming mode.

### Note

Some of the security assumptions of the original Move storage instruction changed when AccountStorage provided Object mode access methods.

For example, `borrow_global<T>`, which is equivalent to `context::borrow_resource<T>`, is constrained by `private_generics<T>`.

But anyone can get `&Object<AccountStorage>`, which means they can get `&AccountStorage`, and can get a read-only reference to any resource via `account::borrow_resource<T>(&AccountStorage):&T`. 

This feature provides more flexibility for state access, i.e., on Rooch, all read-only references to the state are public. However, it is important to be aware of the security issues. In some `Capability` designs, `&Cap` cannot be used as an access control, `&mut Cap` must be used.

### TODO
1. Refactor Framework.
2. Fix tests and examples.

### CODE REVIEW

File: moveos/moveos-stdlib/moveos-stdlib/sources/entrys.move [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+/// Collection of entry functions
+//TODO maybe need a better name
+module moveos_std::entrys {
+    use std::vector;
+    use moveos_std::context::{Self, Context};
+    use moveos_std::move_module::{Self, MoveModule};
+
+    /// Entry function to publish modules
+    /// The order of modules must be sorted by dependency order.
+    public entry fun publish_modules(ctx: &mut Context, account: &signer, modules: vector<vector<u8>>) {
```
Comment (baichuan3) [inline moveos/moveos-stdlib/moveos-stdlib/sources/entrys.move:12]: Are all framework entry functions defined here?

File: moveos/moveos-stdlib/moveos-stdlib/sources/account_storage.move [hunk 2]
```diff
@@ -115,61 +74,37 @@ module moveos_std::account_storage {
...
-    /// Check if the account has a resource of the given type
-    /// This function equates to `exists<T>(address)` instruction in Move
-    public fun global_exists<T: key>(ctx: &Context, account: address) : bool {
-        if (exist_account_storage(ctx, account)) {
-            let account_storage = borrow_account_storage(ctx, account);
-            exists_resource_at_account_storage<T>(account_storage)
-        }else{
-            false
-        }
+    public(friend) fun transfer(obj: Object<AccountStorage>, account: address) {
```
Comment (baichuan3) [inline moveos/moveos-stdlib/moveos-stdlib/sources/account_storage.move:99]: This should be `transfer_extend`?

File: moveos/moveos-stdlib/moveos-stdlib/sources/account_storage.move [hunk 3]
```diff
@@ -115,61 +74,37 @@ module moveos_std::account_storage {
...
+        borrow_resource_from_account_storage<T>(self)
+    }
 
-    #[private_generics(T)]
-    /// Borrow a resource from the account's storage
-    /// This function equates to `borrow_global<T>(address)` instruction in Move
-    public fun global_borrow<T: key>(ctx: &Context, account: address): &T {
-        let account_storage = borrow_account_storage(ctx, account);
-        borrow_resource_from_account_storage<T>(account_storage)
+    public fun borrow_mut_resource<T: key>(self: &mut AccountStorage): &mut T {
```
Comment (baichuan3) [inline moveos/moveos-stdlib/moveos-stdlib/sources/account_storage.move:83]: AbstractStorage has two types of storage, one is resource and the other is object. Do we need to consider naming objects similar to ObjectStorage?
Another question, StorageContext is almost unused in the framework now, will it be abandoned directly in the future?

File: moveos/moveos-stdlib/moveos-stdlib/sources/account_storage.move (hunk 2 above)
Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/account_storage.move:99]: AccountStorage does not support public transfer and does not allow to extend.

File: moveos/moveos-stdlib/moveos-stdlib/sources/account_storage.move (hunk 3 above)
Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/account_storage.move:83]: StorageContext is for supporting App Storage Context in the future. It is unused now and does not affect the Framework API.

File: moveos/moveos-stdlib/moveos-stdlib/sources/context.move [hunk 4]
```diff
@@ -103,8 +103,89 @@ module moveos_std::context {
         tx_context::tx_result(&self.tx_context)
     }
 
+    // === Table functions ===
+
+    public fun new_table<K: copy + drop, V: store>(self: &mut Context): Table<K, V>{
```
[the remark this answers is not in the document]
Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/context.move:108]: `context::new_table` is a wrapper to `table::new`. In most cases, the developer will use the `context::new_table`. If they encounter the two mut reference issues, they can use `table::new` to resolve it.

File: moveos/moveos-stdlib/moveos-stdlib/sources/entrys.move (hunk 1 above)
Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/entrys.move:12]: To resolve the dependency issue, `account_storage` can not depend on `context`. Maybe we can put it into the `package.move` module as @pause125 mentioned.

File: moveos/moveos-stdlib/moveos-stdlib/sources/account_storage.move (hunk 3 above)
Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/account_storage.move:83]: For Object, now all Object storage functions are implemented in the `object` module. Maybe we can migrate it into the `storage_context` or `object_storage` module, but these are all the implementation details. 

Account resource and object public API in the `context`, The refactor does not affect the developer.

File: moveos/moveos-stdlib/moveos-stdlib/sources/context.move [hunk 5]
```diff
@@ -103,8 +103,89 @@ module moveos_std::context {
...
+        if (exist_account_storage(self, account)) {
+            let account_storage = borrow_account_storage(self, account);
+            account_storage::exists_resource<T>(account_storage)
+        }else{
+            false
+        }
+    }
+
+    /// Publish modules to the account's storage
+    public fun publish_modules(self: &mut Context, account: &signer, modules: vector<MoveModule>) {
```
[the remark this answers is not in the document]
Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/context.move:168]: I record an issue #1119. But for now, we need to ensure the `publish_modules` and `publish_modules` entry function locations and our future refactors do not affect the developer.

I prefer to migrate the `publish_modules` entry function into `context`.

[the remark this answers is not in the document]
Comment (jolestar) [reply moveos/moveos-stdlib/moveos-stdlib/sources/context.move:168]: The `entrys.move` is difficult to remember

**Review by baichuan3 — APPROVED** [review]

**Pull request merged**